src/lenskit/data/relationships.py

```diff
@@ -163,19 +184,86 @@ def pandas(self, *, attributes: str | list[str] | None = None, ids=False) -> pd.
+            return new_mat_rset
+
+    def _make_matrix(
+        self, *, row_entity: str = "user", col_entity: str = "item"
     ) -> MatrixRelationshipSet:  # pragma: nocover
         """
         Convert this relationship set into a matrix, coalescing duplicate
         observations.
 
         Args:
-            combine:
-                The method for combining attribute values for repeated
-                relationships or interactions.  Can either be a single strategy
-                or a mapping from attribute names to combination strategies.
+            row_entity:
+                The specified row entity of the matrix
+            col_entity:
+                The specified column entity of the matrix
         """
-        raise NotImplementedError()
+        if row_entity not in self.schema.entities.keys():
+            raise FieldError(self.name, row_entity)
+
+        if col_entity not in self.schema.entities.keys():
+            raise FieldError(self.name, col_entity)
+
+        if col_entity == row_entity:
+            raise ValueError("row and column entity should not be the same")
+
+        e_dict: dict[str, str | None]
+        e_dict = {row_entity: None, col_entity: None}
+        matrix_schema = RelationshipSchema(
+            entities=e_dict,
+            interaction=self.schema.interaction,
+            repeats=self.schema.repeats,
+            attributes=self.schema.attributes,
+            remove_duplicates=self.schema.remove_duplicates,
+        )
+        new_table = self._table
+
+        if "timestamp" in new_table.column_names:
+            new_table = new_table.sort_by([("timestamp", "ascending")])
+
+        if "count" not in new_table.column_names:
+            new_table = new_table.add_column(
+                new_table.num_columns, "count", pa.array(np.ones(new_table.num_rows))
+            )
+        else:
+            null_filled = new_table["count"].fill_null(1)
+            new_table = new_table.set_column(
+                new_table.schema.get_field_index("count"),
+                "count",
+                null_filled,  # type: ignore
+            )
+
+        group_keys = [entity + "_num" for entity in e_dict]
+        table_group = new_table.group_by(group_keys)
+
+        aggregates: list[tuple[str, str]]
+        column_renames: dict[str, str]
+        aggregates = []
+        column_renames = {}
+
+        if "timestamp" in new_table.column_names:
+            aggregates.extend([("timestamp", "max"), ("timestamp", "min")])
+            column_renames.update(
+                {"timestamp_max": "timestamp", "timestamp_min": "first_timestamp"}
+            )
+
+        if "count" in new_table.column_names:
+            aggregates.append(("count", "sum"))
+            column_renames.update({"count_sum": "count"})
+
+        if aggregates:
+            aggregated_table = table_group.aggregate(aggregates)
+            aggregated_table = aggregated_table.rename_columns(column_renames)
+
```

Comment:
A point we forgot: for each of the other columns (not `timestamp`, `count`, or one of the entity number columns), we need an aggregate that takes the `last` value (since we already sort by timestamp). If we have any such aggregates, then we also need to pass `use_threads=False` to `table_group.aggregate`, because paralllel aggregation doesn't work with the `last` aggregate.